Add XCCL backend for distributed recipes

Kanya-Mo opened the pull request:
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature

XCCL as a torch distributed backend has been enabled for XPU device. The PR enabled it for distributed recipes.

#### Changelog
Added XCCL backend option and turned on XPU memory stats logging in distributed training.

#### Test plan
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)

#### UX
- [x] I did not change any public API

pbontrager commented:
Thank you for putting up this PR! I looked through it and it seems you're fixing the backend option for recipes that still use hardcoded backends which we haven't updated yet. Can you instead update the function `torchtune/training/_distributed.py:get_distributed_backend` and then update the recipes you edited to look like [this](https://github.com/pytorch/torchtune/blob/8a10fad5bf67adffe0c0d3a391b240a532116b07/recipes/full_finetune_distributed.py#L135C14-L135C33) instead? Thank you again!

Kanya-Mo commented:
> Thank you for putting up this PR! I looked through it and it seems you're fixing the backend option for recipes that still use hardcoded backends which we haven't updated yet. Can you instead update the function `torchtune/training/_distributed.py:get_distributed_backend` to support xccl and then update the recipes you edited to look like [this](https://github.com/pytorch/torchtune/blob/8a10fad5bf67adffe0c0d3a391b240a532116b07/recipes/full_finetune_distributed.py#L135C14-L135C33) instead? Thank you again!

Sure, I have updated the way these recipes set up backend as the example and verified the function does support XCCL. `recipes/full_dpo_distributed.py` and `recipes/lora_dpo_distributed.py` need to sync on cpu tensors (some metrics) as well, so I fixed `offload_ops_to_cpu` in these two recipes to include gloo backend.

joecummings commented on recipes/dev/lora_finetune_distributed_multi_dataset.py:
```diff
@@ -146,9 +156,9 @@ def __init__(self, cfg: DictConfig) -> None:
         self._log_peak_memory_stats = cfg.get("log_peak_memory_stats", False)
 
-        if self._log_peak_memory_stats and self._device.type != "cuda":
+        if self._log_peak_memory_stats and self._device.type == "cpu":
```
Can you change this to `self._device.type in {"cuda", "xpu"}` rather than == cpu?

It's a more accurate check for log_peak_memory steps.

joecummings reviewed (commented):
Looks great! Please attach some proof of training running (logs are fine) and let's get this landed :)

joecummings reviewed (approved):
Great work, thank you! @Kanya-Mo

Kanya-Mo commented:
Attached are the logs from some example recipes on XPU

- tune run --nnodes 1 --nproc_per_node 4 full_finetune_distributed --config llama2/7B_full max_steps_per_epoch=10 device=xpu

```
Step 1 | loss:2.17120361328125 lr:2e-05 tokens_per_second_per_gpu:8.861910820007324 peak_memory_active:13.46799898147583 peak_memory_alloc:12.595917224884033 peak_memory_reserved:19.029296875
Step 2 | loss:1.5850075483322144 lr:2e-05 tokens_per_second_per_gpu:99.43350982666016 peak_memory_active:16.83338165283203 peak_memory_alloc:14.558853149414062 peak_memory_reserved:22.662109375
Step 3 | loss:1.1229164600372314 lr:2e-05 tokens_per_second_per_gpu:177.71878051757812 peak_memory_active:26.27274227142334 peak_memory_alloc:14.558753967285156 peak_memory_reserved:28.7265625
Step 4 | loss:0.8744106292724609 lr:2e-05 tokens_per_second_per_gpu:112.53812408447266 peak_memory_active:15.685035228729248 peak_memory_alloc:14.55413007736206 peak_memory_reserved:28.7265625
Step 5 | loss:0.8242135047912598 lr:2e-05 tokens_per_second_per_gpu:144.79156494140625 peak_memory_active:15.299323081970215 peak_memory_alloc:14.54538631439209 peak_memory_reserved:28.7265625
Step 6 | loss:0.9700961112976074 lr:2e-05 tokens_per_second_per_gpu:187.36773681640625 peak_memory_active:17.698896884918213 peak_memory_alloc:14.594339847564697 peak_memory_reserved:28.73046875
Step 7 | loss:0.7872267365455627 lr:2e-05 tokens_per_second_per_gpu:149.49046325683594 peak_memory_active:17.94667387008667 peak_memory_alloc:14.548095226287842 peak_memory_reserved:28.73046875
Step 8 | loss:0.9488868713378906 lr:2e-05 tokens_per_second_per_gpu:167.1237335205078 peak_memory_active:19.540390491485596 peak_memory_alloc:14.639801502227783 peak_memory_reserved:29.111328125
Step 9 | loss:0.9222078323364258 lr:2e-05 tokens_per_second_per_gpu:186.1638641357422 peak_memory_active:28.880515575408936 peak_memory_alloc:14.55571699142456 peak_memory_reserved:30.626953125
Step 10 | loss:0.8577423095703125 lr:2e-05 tokens_per_second_per_gpu:166.01620483398438 peak_memory_active:21.339026927947998 peak_memory_alloc:14.553596019744873 peak_memory_reserved:30.626953125

```

- tune run --nnodes 1 --nproc_per_node 2 lora_finetune_distributed --config llama2/7B_lora max_steps_per_epoch=10 device=xpu
```
Step 1 | loss:1.4962120056152344 lr:2.9999999999999997e-06 tokens_per_second_per_gpu:68.69140625 peak_memory_active:19.47888994216919 peak_memory_alloc:17.60053062438965 peak_memory_reserved:22.779296875
Step 2 | loss:1.5554665327072144 lr:5.999999999999999e-06 tokens_per_second_per_gpu:559.3739624023438 peak_memory_active:19.92483377456665 peak_memory_alloc:17.270742893218994 peak_memory_reserved:24.30859375
Step 3 | loss:1.5655124187469482 lr:8.999999999999999e-06 tokens_per_second_per_gpu:700.7313842773438 peak_memory_active:20.280246257781982 peak_memory_alloc:18.77446460723877 peak_memory_reserved:27.89453125
Step 4 | loss:1.5661518573760986 lr:1.1999999999999999e-05 tokens_per_second_per_gpu:536.4591064453125 peak_memory_active:19.590736865997314 peak_memory_alloc:16.294240951538086 peak_memory_reserved:27.89453125
Step 5 | loss:1.3553907871246338 lr:1.4999999999999999e-05 tokens_per_second_per_gpu:783.439208984375 peak_memory_active:24.94785737991333 peak_memory_alloc:20.046388626098633 peak_memory_reserved:33.21484375
Step 6 | loss:1.5427088737487793 lr:1.7999999999999997e-05 tokens_per_second_per_gpu:753.8826904296875 peak_memory_active:16.782937049865723 peak_memory_alloc:14.256525993347168 peak_memory_reserved:33.216796875
Step 7 | loss:1.53976309299469 lr:2.1e-05 tokens_per_second_per_gpu:801.1482543945312 peak_memory_active:22.29063653945923 peak_memory_alloc:17.75285768508911 peak_memory_reserved:33.595703125
Step 8 | loss:1.6075184345245361 lr:2.3999999999999997e-05 tokens_per_second_per_gpu:715.87451171875 peak_memory_active:21.96659755706787 peak_memory_alloc:18.55644130706787 peak_memory_reserved:34.734375
Step 9 | loss:1.4994750022888184 lr:2.6999999999999996e-05 tokens_per_second_per_gpu:720.1342163085938 peak_memory_active:23.5574893951416 peak_memory_alloc:19.06113862991333 peak_memory_reserved:34.734375
Step 10 | loss:1.5589946508407593 lr:2.9999999999999997e-05 tokens_per_second_per_gpu:791.6353759765625 peak_memory_active:25.752007961273193 peak_memory_alloc:18.59989833831787 peak_memory_reserved:36.25390625
```

- tune run --nnodes 1 --nproc_per_node 2 lora_dpo_distributed --config llama2/7B_lora_dpo max_steps_per_epoch=10 device=xpu
```
Step 1 | loss:1.388274908065796 lr:5e-06 tokens_per_second_per_gpu:596.296875 rewards/chosen:0.002933537820354104 rewards/rejected:0.004445761442184448 rewards/accuracies:0.484375 rewards/margins:-0.0015122236218303442 log_probs/chosen:-334.84991455078125 log_probs/rejected:-313.72991943359375 logits/chosen:-2.5304861068725586 logits/rejected:-2.785830020904541 peak_memory_active:23.43517303466797 peak_memory_alloc:14.303722858428955 peak_memory_reserved:27.01953125
Step 2 | loss:1.394402265548706 lr:1e-05 tokens_per_second_per_gpu:555.7238159179688 rewards/chosen:0.002613967750221491 rewards/rejected:0.010205138474702835 rewards/accuracies:0.453125 rewards/margins:-0.007591170724481344 log_probs/chosen:-311.7193908691406 log_probs/rejected:-237.09597778320312 logits/chosen:-2.1929900646209717 logits/rejected:-2.528794765472412 peak_memory_active:24.073253631591797 peak_memory_alloc:14.336819171905518 peak_memory_reserved:27.0234375
Step 3 | loss:1.3852393627166748 lr:1.5e-05 tokens_per_second_per_gpu:723.3158569335938 rewards/chosen:-0.004834341816604137 rewards/rejected:-0.006729197688400745 rewards/accuracies:0.421875 rewards/margins:0.001894855871796608 log_probs/chosen:-377.57696533203125 log_probs/rejected:-333.8240966796875 logits/chosen:-2.208315372467041 logits/rejected:-2.5521175861358643 peak_memory_active:24.144123077392578 peak_memory_alloc:14.337162494659424 peak_memory_reserved:27.40234375
Step 4 | loss:1.3799787759780884 lr:2e-05 tokens_per_second_per_gpu:840.6644287109375 rewards/chosen:0.0012696684570983052 rewards/rejected:-0.00579102011397481 rewards/accuracies:0.5 rewards/margins:0.007060688454657793 log_probs/chosen:-369.2913818359375 log_probs/rejected:-300.900146484375 logits/chosen:-2.0673460960388184 logits/rejected:-2.588378429412842 peak_memory_active:24.139354705810547 peak_memory_alloc:14.336819171905518 peak_memory_reserved:27.40234375
Step 5 | loss:1.3865116834640503 lr:2.5e-05 tokens_per_second_per_gpu:869.0836791992188 rewards/chosen:0.00016164779663085938 rewards/rejected:-0.0002714156871661544 rewards/accuracies:0.5 rewards/margins:0.00043306348379701376 log_probs/chosen:-319.62060546875 log_probs/rejected:-308.67742919921875 logits/chosen:-2.558422803878784 logits/rejected:-2.600210428237915 peak_memory_active:24.07415008544922 peak_memory_alloc:14.336819171905518 peak_memory_reserved:27.40234375
Step 6 | loss:1.373827338218689 lr:3e-05 tokens_per_second_per_gpu:795.5487670898438 rewards/chosen:0.01076948456466198 rewards/rejected:-0.00249136658385396 rewards/accuracies:0.515625 rewards/margins:0.013260850682854652 log_probs/chosen:-349.17120361328125 log_probs/rejected:-256.47015380859375 logits/chosen:-1.8775806427001953 logits/rejected:-2.216588020324707 peak_memory_active:24.413097381591797 peak_memory_alloc:14.336857318878174 peak_memory_reserved:27.40234375
Step 7 | loss:1.385218858718872 lr:3.5000000000000004e-05 tokens_per_second_per_gpu:876.3935546875 rewards/chosen:0.007726062089204788 rewards/rejected:0.005846577696502209 rewards/accuracies:0.546875 rewards/margins:0.0018794843927025795 log_probs/chosen:-421.33819580078125 log_probs/rejected:-322.2149658203125 logits/chosen:-2.12746524810791 logits/rejected:-2.5231056213378906 peak_memory_active:24.452159881591797 peak_memory_alloc:14.336819171905518 peak_memory_reserved:27.40234375
Step 8 | loss:1.3796520233154297 lr:4e-05 tokens_per_second_per_gpu:787.59814453125 rewards/chosen:0.003570080269128084 rewards/rejected:-0.0037251829635351896 rewards/accuracies:0.515625 rewards/margins:0.00729526299983263 log_probs/chosen:-263.75677490234375 log_probs/rejected:-278.4831848144531 logits/chosen:-2.319993495941162 logits/rejected:-2.233083963394165 peak_memory_active:24.372051239013672 peak_memory_alloc:14.336819171905518 peak_memory_reserved:27.40234375
Step 9 | loss:1.3904856443405151 lr:4.4999999999999996e-05 tokens_per_second_per_gpu:815.1632690429688 rewards/chosen:-0.0012891886290162802 rewards/rejected:0.002246975665912032 rewards/accuracies:0.46875 rewards/margins:-0.0035361642949283123 log_probs/chosen:-379.044921875 log_probs/rejected:-307.0588684082031 logits/chosen:-2.2394227981567383 logits/rejected:-2.6336183547973633 peak_memory_active:24.24807357788086 peak_memory_alloc:14.336819171905518 peak_memory_reserved:27.40234375
Step 10 | loss:1.3888099193572998 lr:5e-05 tokens_per_second_per_gpu:662.84521484375 rewards/chosen:0.010051174089312553 rewards/rejected:0.011655766516923904 rewards/accuracies:0.515625 rewards/margins:-0.001604592427611351 log_probs/chosen:-365.03204345703125 log_probs/rejected:-285.5168762207031 logits/chosen:-2.346580743789673 logits/rejected:-2.795206069946289 peak_memory_active:24.43549346923828 peak_memory_alloc:14.336819171905518 peak_memory_reserved:27.40234375
```

- tune run --nnodes 1 --nproc_per_node 2 knowledge_distillation_distributed --config llama3_2/8B_to_1B_KD_lora_distributed max_steps_per_epoch=10 device=xpu

```
Step 1 | loss:2.0439305901527405 class_loss:2.2051808834075928 kd_loss:1.8826802968978882 lr:2.9999999999999997e-06 tokens_per_second_per_gpu:149.5618896484375
Step 2 | loss:1.5403059720993042 class_loss:1.6840832233428955 kd_loss:1.396528720855713 lr:5.999999999999999e-06 tokens_per_second_per_gpu:311.7116394042969
Step 3 | loss:2.153505563735962 class_loss:2.2536041736602783 kd_loss:2.0534069538116455 lr:8.999999999999999e-06 tokens_per_second_per_gpu:348.604248046875
Step 4 | loss:1.8863587975502014 class_loss:2.0781383514404297 kd_loss:1.6945792436599731 lr:1.1999999999999999e-05 tokens_per_second_per_gpu:330.8072814941406
Step 5 | loss:2.026710629463196 class_loss:2.1081862449645996 kd_loss:1.945235013961792 lr:1.4999999999999999e-05 tokens_per_second_per_gpu:331.4019470214844
Step 6 | loss:2.1739938259124756 class_loss:2.329331874847412 kd_loss:2.018655776977539 lr:1.7999999999999997e-05 tokens_per_second_per_gpu:264.285888671875
Step 7 | loss:1.682055413722992 class_loss:1.8389090299606323 kd_loss:1.5252017974853516 lr:2.1e-05 tokens_per_second_per_gpu:323.925537109375
Step 8 | loss:1.85639226436615 class_loss:1.9893798828125 kd_loss:1.7234046459197998 lr:2.3999999999999997e-05 tokens_per_second_per_gpu:351.9089050292969
Step 9 | loss:2.118502140045166 class_loss:2.205515146255493 kd_loss:2.031489133834839 lr:2.6999999999999996e-05 tokens_per_second_per_gpu:473.61004638671875
Step 10 | loss:1.8005318641662598 class_loss:1.8205370903015137 kd_loss:1.7805266380310059 lr:2.9999999999999997e-05 tokens_per_second_per_gpu:474.53521728515625
```
